---
title: "DE DAS and DTU analysis"
header-includes: \usepackage{caption}
output:
  html_document:
    fig_caption: yes
    highlight: textmate
    theme: default
    toc: yes
    toc_depth: 4
    toc_float:
      collapsed: no
      smooth_scroll: yes
  word_document:
    fig_caption: yes
    toc: yes
    toc_depth: '4'
  pdf_document:
    fig_caption: yes
    toc: yes
    toc_depth: '4'
date: "`r format(Sys.time(), '%d %B, %Y')`"
params:
  DDD.data: !r list()
  # DDD.data: !r list()
  # DDD.data$params_list: !r list()
  # DAS_pval_method: 'F-test'
  # DDD_numbers: !r data.frame()
  # DEvsDAS_results: !r data.frame()
  # DEvsDTU_results: !r data.frame()
  # RNAseq_info: !r data.frame()
  # DDD.data: !r list()
  # RUVseq_method: 'RUVr'
  # brep_n: 3
  # condition_n: 26
  # contrast: !r c('T10-T2','T19-T2')
  # cpm_cut: 1
  # cpm_samples_n: 3
  # deltaPS_cut: 0.1
  # has_batcheffect: 'Yes'
  # has_srep: 'Yes'
  # l2fc_cut: 1
  # norm_method: 'TMM'
  # pval_adj_method: 'BH'
  # pval_cut: 0.01
  # quant_method: 'salmon'
  # samples_n: 234
  # srep_n: 3
  # tximport_method: 'lengthScaledTPM'
  # DE_genes: !r data.frame()
  # DAS_genes: !r data.frame()
  # target_high: !r list()
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = F,eval = T,error = T)
```

<div align="justify">
To use our pipeline in your work, please cite:

-   Guo,W., Tzioutziou,N., Stephen,G., Milne,I., Calixto,C., Waugh,R., Brown,J.W., and Zhang,R. (2019) 3D RNA-seq - a powerful and flexible tool for rapid and accurate differential expression and alternative splicing analysis of RNA-seq data for biologists. bioRxiv, 656686. doi: https://doi.org/10.1101/656686.
-   Calixto,C.P.G., Guo,W., James,A.B., Tzioutziou,N.A., Entizne,J.C., Panter,P.E., Knight,H., Nimmo,H.G., Zhang,R., and Brown,J.W.S. (2018) Rapid and Dynamic Alternative Splicing Impacts the Arabidopsis Cold Response Transcriptome. Plant Cell, 30, 1424–1444.


## Method
The 3D RNA-seq App was developed for rapid and accurate differential expression (DE), differential alternative splicing (DAS) gene and differential transcript usage (DTU) (3D) analysis (Guo et al., 2019; Calixto et al., 2018).

### RNA-seq datasets
The RNA-seq data had `r params$DDD.data$params_list$condition_n` factor groups (`r paste0(unique(params$DDD.data$conditions),collapse = ', ')`) and each had `r params$DDD.data$params_list$brep_n` biological replicates `r ifelse(params$DDD.data$params_list$has_srep=='Yes',paste0(' and ',params$DDD.data$params_list$srep_n,' sequencing replicates'),'')` (`r params$DDD.data$params_list$samples_n` samples in total).

```{r experimental_design,eval=T,echo=F,fig.width=12}
fig2insert <- paste0(DDD.data$figure.folder,'/experimental_design.png')
if(file.exists(fig2insert)){
  knitr::include_graphics(fig2insert)
}

```

```{r,eval=T,echo=F,results='asis'}
fig2insert <- paste0(DDD.data$figure.folder,'/experimental_design.png')
if(file.exists(fig2insert)){
  cat('**Figure**: Experimental design.')
}

```

### Data  pre-processing
Read counts and transcript per million reads (TPMs) were generated using tximport R package version 1.10.0 and `r params$DDD.data$params_list$tximport_method` method (Soneson et al., 2016) with inputs of transcript quantifications from tool `r params$DDD.data$params_list$quant_method` `r ifelse(params$DDD.data$params_list$quant_method=='salmon', '(Patro et al., 2017)','(Bray et al., 2016)')`. `r ifelse(params$DDD.data$params_list$has_srep=='Yes',paste0('Sequencing replicates were merged to increase the sequencing depth, giving ',params$DDD.data$params_list$condition_n*params$DDD.data$params_list$brep_n,' samples.'),'')` Low expressed transcripts and genes were filtered based on analysing the data mean-variance trend. The expected decreasing trend between data mean and variance was observed when expressed transcripts were determined as which had $\geq$ `r params$DDD.data$params_list$cpm_samples_n` of the `r params$DDD.data$params_list$condition_n*params$DDD.data$params_list$brep_n` samples with count per million reads (CPM) $\geq$ `r params$DDD.data$params_list$cpm_cut`, which provided an optimal filter of low expression. A gene was expressed if any of its transcripts with the above criteria was expressed. The `r params$DDD.data$params_list$norm_method` method was used to normalise the gene and transcript read counts to $log_2$-CPM (Bullard et al., 2010). The principal component analysis (PCA) plot showed the RNA-seq data `r ifelse(params$DDD.data$params_list$has_batcheffect=='Yes',paste0('had distinct batch effects between biological replicates. Thus, RUVSeq R package version 1.16.0 with approach ',params$DDD.data$params_list$RUVseq_method,' was used to estimate the batch effects (Risso et al., 2014).'),' did not have distinct batch effects. Downstream analysis can be directly proceeded.')`

### DE, DAS and DTU analysis
`r ifelse(params$DDD.data$params_list$DE_pipeline=='limma', 'Limma R package was used for 3D expression comparison (Ritchie et al., 2015; Law et al., 2014).',paste0('The ',params$DDD.data$params_list$DE_pipeline,' method of edgeR R package was used for 3D expression comparison (Robinson et al., 2011).'))` To compare the expression changes between conditions of experimental design, the contrast groups were set as `r paste0(params$DDD.data$params_list$contrast,collapse = ', ')`. For DE genes/transcripts, the $log_2$ fold change ($L_2FC$) of gene/transcript abundance were calculated based on contrast groups and significance of expression changes were determined using t-test (Figure 1). P-values of multiple testing were adjusted with `r params$DDD.data$params_list$pval_adj_method` to correct false discovery rate (FDR) (Benjamini and Yekutieli, 2001). A gene/transcript was significantly DE in a contrast group if it had adjusted p-value < `r params$DDD.data$params_list$pval_cut` and $L_2FC$ $\geq$ `r params$DDD.data$params_list$l2fc_cut`.

At the alternative splicing level, DTU transcripts were determined by comparing the $L_2FC$ of a transcript to the `r ifelse(params$DDD.data$params_list$DE_pipeline=='limma','weighted average','sum')` of $L_2FCs$ `r ifelse(params$DDD.data$params_list$DE_pipeline=='limma','(weights were based on their standard deviation)','')` of all `r ifelse(params$DDD.data$params_list$DE_pipeline=='limma','remaining','')` transcripts in the same gene. A transcript was determined as significant DTU if it had adjusted p-value < `r params$DDD.data$params_list$pval_cut` and $\Delta$PS $\geq$ `r params$DDD.data$params_list$deltaPS_cut` (Figure 1). For DAS genes, `r ifelse(params$DDD.data$params_list$DE_pipeline=='limma','each individual transcript $L_2FC$ were compared to gene level $L_2FC$, which was calculated as the weighted average of $L_2FCs$ of all transcripts of the gene. Then','the')` p-values of individual transcript comparison were summarised to a single gene level p-value with `r params$DDD.data$params_list$DAS_pval_method`. A gene was significantly DAS in a contrast group if it had an adjusted p-value < `r params$DDD.data$params_list$pval_cut` and any of its transcript had a $\Delta$ Percent Spliced ($\Delta$PS) ratio $\geq$ `r params$DDD.data$params_list$deltaPS_cut` (Figure 1). 

<!-- <img style="width: 100%; display: block; margin-left: auto; margin-right: auto;" src="www/DDD_test.png"/> -->

![](www/DDD_test.png)

**Figure 1**: Testing of DE genes, DAS genes, DE transcripts and DTU transcripts.

```{r,echo=FALSE}
whether_tstrend <- params$DDD.data$params_list$whether_tstrend =='Yes'
```

```{r conditional_block_tstrend_method, echo=FALSE, results='asis', eval=whether_tstrend}
cat('### Time-series trend analysis\n')
cat('Time-points (or equivalence, e.g. developmental series and conditions with sequential orders):',paste0(levels(params$DDD.data$tstrend_timePoints),collapse = ', '), 'in groups: ',paste0(levels(params$DDD.data$tstrend_timeGroups),collapse = ', '), 'are used for 3D trend analysis (see the schematic diagram).\n')
cat(ifelse(params$DDD.data$tstrend_spline=='Yes',paste0('Natural Cubic Spline method with degree of freedom ',params$DDD.data$tstrend_spline_df,' was used to generate time-series trend.\n'),''))
cat('Contrast groups of trend were set as:',paste0(paste0(names(params$DDD.data$tstrend_Contrastgroups),collapse = ', '),'.\n'))
cat('Significant 3D time-series trend genes and transcripts were determined by',params$DDD.data$params_list$pval_adj_method_tstrend,'adjusted p-value <',paste0(params$DDD.data$params_list$pval_cut_tstrend,'.\n'))
cat('At alternative splicing level, the p-value of a DAS gene was summarised from the p-values of transcript level DTU analysis with',params$DDD.data$params_list$DAS_pval_method_tstrend,'method.\n')
cat('The overall DAS gene/DTU transcript p-values across time groups in each of the contrast groups were summarised by',params$DDD.data$params_list$DAS_pval_method_across_group,'method.\n')

```

```{r,echo=FALSE,fig.height = 4, fig.width = 9, fig.align = "center", eval=whether_tstrend}
# Time0 <- Time.points()
# Time <- as.numeric(factor(Time0))
Time.points <- params$DDD.data$tstrend_timePoints
Time.groups <- params$DDD.data$tstrend_timeGroups
Time <- as.numeric(factor(Time.points))
set.seed(100)
Expression <- sin(Time*pi/6)+rnorm(length(Time),1,0.5)
Expression <- Expression*as.numeric(factor(Time.groups))

if(params$DDD.data$tstrend_spline == 'Yes'){
  Time.ns <- splines::ns(Time,df=params$DDD.data$tstrend_spline_df)
  knots <- sort(c(attr(Time.ns,'knots'),attr(Time.ns,'Boundary.knots')),decreasing = F)
  data2fit <- data.frame(Time=Time,Expression=Expression,Group=Time.groups)
  data2line <- by(data = data2fit,INDICES = data2fit$Group,function(x){
    fit <- tryCatch({lm(Expression ~ splines::ns(Time,df=params$DDD.data$tstrend_spline_df),data = x)},
                    error=function(e){})
    if(is.null(fit))
      return(NULL)
    pred.value <- predict(fit,data.frame(Time=knots))
    names(pred.value) <- knots
    pred.value
  })
  data2line <- do.call(rbind,data2line)

  data2line <- reshape2::melt(data2line)
  colnames(data2line) <- c('Group','Time','Expression')
  
  data2plot <- data.frame(Time=Time,Label=Time.points,Group=Time.groups,Expression=Expression)
  # data2line <- data.frame(Time=knots,Expression=pred.value)
  g <- ggplot(data = data2plot,aes(x=Time,y=Expression))+
    geom_point(aes(pch='Samples reps'),colour='black')+
    scale_shape_manual(values = c('Samples reps'=1))+
    theme_bw()+
    labs(title='Schematic diagram: spline time-series with Natural Cubic method',shape='Data points')
  q <- g+geom_line(data = data2line,aes(x=Time,y=Expression,colour='red'))+
    geom_point(data = data2line,pch=16,size=3,aes(colour='blue'))+facet_grid(.~Group)+
    scale_color_identity(name = "Spline fit",
                         breaks = c( "red", "blue"),
                         labels = c("Spline fitted line", "Spline knots"),
                         guide = "legend")
} else {
  data2plot <- data.frame(Time=Time.points,Expression=Expression,Group=Time.groups)
  q <- ggplot(data2plot, aes(x=Time, y=Expression)) + 
    geom_point(aes(pch='Samples reps'),colour='black') +
    stat_summary(aes(y = Expression,group=1,colour="Mean value"), fun.y=mean, geom="line",group=1)+
    facet_grid(.~Group)+
    theme_bw()+
    labs(title='Schematic diagram: Time-series trend',shape='Data points',colour='Fitted line')+
    scale_shape_manual(values = c('Samples reps'=1))
}
q <- q+labs(y='Simulated expression')
print(q)

```

### Isoform switch analysis
Transcript isoform switches (ISs) occur when a pair of alternatively spliced isoforms reverse the order of their relative expression levels (Figure 2). In this analysis, `r ifelse(params$DDD.data$params_list$TSISorisokTSP=='isokTSP',paste0('Pair-Wise Isoform Switch (isokTSP) method was used to detect the isoform switch points between conditions of contrasts groups: ',paste0(params$DDD.data$contrast_pw,collapse = ', '), ' (Figure 2A)'),' Time-Series Isoform switch (TSIS) method was used to detect the isoform switch points along the sequencial time-points (Figure 2B).')` (Guo et al., 2017). The `r length(which(params$DDD.data$target_high$mapping_high$GENEID %in% unique(params$DDD.data$DAS_genes$target)))` expressed transcripts  of `r length(unique(params$DDD.data$DAS_genes$target))` DAS genes were used for the analysis. The method defined the ISs between any pair of transcripts within genes using `r params$DDD.data$params_list$TSIS_method_intersection` values of conditions. It described the significant ISs using five different features of metrics: 1) the probability of switch (i.e. the frequency of samples reversing their relative abundance at the switches) was set to > `r params$DDD.data$params_list$TSIS_prob_cut`; (2)  the sum of the average differences of the two isoforms in both intervals before and after the switch point were set at $\Delta$TPM > `r params$DDD.data$params_list$TSIS_diff_cut`; (3) the significance of the differences between the switched isoform abundances before and after the switch was set to BH adjusted p-value < `r params$DDD.data$params_list$TSIS_adj_pval_cut`; (4) both of the interval lengths before and after switch were set to `r params$DDD.data$params_list$TSIS_time_point_cut`; (5) Pearson correlation of two isoforms was set to >`r params$DDD.data$params_list$TSIS_cor_cut` (see the paper Guo et al., (2017) for methodology details).

<!-- <img style="width: 100%; display: block; margin-left: auto; margin-right: auto;" src="www/TSIS.png"/> -->

![](www/TSIS.png)

```{r,eval=T}
getwd()

list.files('www',full.names = T)
```

**Figure 2**: Isoform switch analysis methods. Expression data with 3 replicates for each condition/time-point is simulated for isoforms $iso_{i}$ and $iso_j$ (blue and red circles). The points in the plots represent data samples and the black lines connect the average of samples. (A) is the Pair-Wise Isoform Switch (isokTSP) method for comparisons of two conditions $c_1$ and $c_2$ (e.g. conditions in contrast groups of 3D RNA-seq analysis). The Time-Series Isoform Switch (TSIS) tool is designed for detection and characterization of isoform switches for time series data shown in (B). The time-series with 6 time-points is divided into 4 intervals by the intersection points of average expression. If the conditions/time-points on x-axis are not numeric, they will be converted to numeric coordinates 1, 2, 3, ... to fit the lines.

## Results
### RNA-seq data variation
Average expression of transcript and gene level log2-CPMs `r ifelse(params$DDD.data$params_list$has_batcheffect=='Yes','(batch effects were removed)','')` were used to make the Principal Component Analysis (PCA) plot to provide visualisation of RNA-seq data variation between conditions of interest.


Figure: Transcript level PCA plot of average expression`r ifelse(params$DDD.data$params_list$has_batcheffect=='Yes',' (after remove batch effects).','.')`

```{r,eval=T,echo=F,fig.width=12,message=F,warning=F}
if(params$DDD.data$params_list$has_batcheffect=='Yes'){
   fig2insert <- paste0(DDD.data$figure.folder,'/Transcript PCA batch effect removed average expression.png')
} else {
   fig2insert <- paste0(DDD.data$figure.folder,'/Transcript PCA average expression.png')
}

if(!file.exists(fig2insert)){
  cat('Plot is not found in the figure folder')
} else {
  knitr::include_graphics(fig2insert)
}
``` 


Figure: Gene level PCA plot of average expression`r ifelse(params$DDD.data$params_list$has_batcheffect=='Yes',' (after remove batch effects).','.')`

```{r,eval=T,echo=F,fig.width=12,message=F,warning=F}
if(params$DDD.data$params_list$has_batcheffect=='Yes'){
   fig2insert <- paste0(DDD.data$figure.folder,'/Gene PCA batch effect removed average expression.png')
} else {
   fig2insert <- paste0(DDD.data$figure.folder,'/Gene PCA average expression.png')
}
if(!file.exists(fig2insert)){
  cat('Plot is not found in the figure folder')
} else {
  knitr::include_graphics(fig2insert)
}
``` 

### Number of transcripts per gene

```{r number_tpg,eval=T,echo=F,fig.width=12}
fig2insert <- paste0(DDD.data$figure.folder,'/Distribution of the number of transcripts per gene.png')
if(!file.exists(fig2insert)){
  cat('Plot is not found in the figure folder')
} else {
  knitr::include_graphics(fig2insert)
}
```

Figure: Distribution of the number of transcripts per gene. Low expressed transcripts and genes are filtered.

### Number of 3D genes and transcripts

Table: RNA-seq sample information before and after data pre-processing. 

```{r Sample_information,eval=T,echo=F}
if(is.null(params$DDD.data$RNAseq_info)){
 cat('Results are not found.')
} else {
  knitr::kable(params$DDD.data$RNAseq_info,format = "markdown",align = 'l')
}
```

<br>

Table: Number of DE/DAS genes and DE/DTU transcripts in different contrast groups. 

```{r 3D_number,eval=T,echo=F}
if(is.null(params$DDD.data$DDD_numbers)){
  cat('Results are not found.')
} else {
  knitr::kable(params$DDD.data$DDD_numbers,format = "markdown",align = 'l')
}
```

<br>

Table: Number of DE vs DAS genes.

```{r DEvsDAS,eval=T,echo=F}
if(is.null(params$DDD.data$DDD_numbers)){
  cat('Results are not found.')
} else {
  knitr::kable(params$DDD.data$DEvsDAS_results,format = "markdown",align = 'l')
}
```

<br>

Table: Number of DE vs DTU transcripts. 

```{r DEvsDTU,eval=T,echo=F}
if(is.null(params$DDD.data$DDD_numbers)){
  cat('Results are not found.')
} else {
  knitr::kable(params$DDD.data$DEvsDTU_results,format = "markdown",align = 'l')
}
```

<br>

```{r Union_set_DEvsDAS,eval=T,echo=F,fig.width=12}
fig2insert <- paste0(DDD.data$figure.folder,'/Union set DE genes vs DAS genes.png')
if(!file.exists(fig2insert)){
  cat('Plot is not found in the figure folder')
} else {
  knitr::include_graphics(fig2insert)
}
```

Figure: Number of genes regulated only by transcription (DE), only by alternative splicing (DAS) and by both transcription and alternative splicing (DE+DAS).

<br>

```{r Union_set_DEvsDTU,eval=T,echo=F,fig.width=12}
fig2insert <- paste0(DDD.data$figure.folder,'/Union set DE transcripts vs DTU transcripts.png')
if(!file.exists(fig2insert)){
  cat('Plot is not found in the figure folder')
} else {
  knitr::include_graphics(fig2insert)
} 

```

Figure: Number of transcripts regulated only by transcription (DE), only by alternative splicing (DAS) and by both transcription and alternative splicing (DE+DAS).



```{r conditional_block_tstrend_results, echo=FALSE, results='asis', eval=whether_tstrend}
cat('### Number of 3D time-series trend genes and transcripts\n')
```

```{r DDD_tstrend_numbers,eval=T,echo=F}
if(is.null(params$DDD.data$DDD_tstrend_numbers)){
  cat('Results are not found.')
} else {
  knitr::kable(params$DDD.data$DDD_tstrend_numbers,format = "markdown",align = 'l')
}
```

### Up- and down-regulation

Figure: Number of up and down regulated genes, DAS genes, DE transcripts and DTU transcripts. The numbers are calculated based on positive or negative signs on the $L_2FCs$ of DE/DAS genes and DE/DTU transcripts. 

```{r updown_DEgenes,eval=T,echo=F,fig.width=12}
fig2insert <- paste0(DDD.data$figure.folder,'/DE genes up and down regulation numbers.png')
if(!file.exists(fig2insert)){
  cat('Plot is not found in the figure folder')
} else {
  knitr::include_graphics(fig2insert)
}
```


```{r updown_DEtrans,eval=T,echo=F,fig.width=12}
fig2insert <- paste0(DDD.data$figure.folder,'/DE transcripts up and down regulation numbers.png')
if(!file.exists(fig2insert)){
  cat('Plot is not found in the figure folder')
} else {
  knitr::include_graphics(fig2insert)
}
```

### Volcano plot
Figure: Volcano plot. The low expressed genes and transcripts were filtered. DE genes: log2FC vs -log10(FDR) at gene level; DAS genes: maximum $\Delta PS$ of transcript in a gene vs -log10(FDR) at gene level; DE transcripts: log2FC vs -log10(FDR) at gene level at transcript level and DTU transcripts: $\Delta PS$ vs -log10(FDR) at transcript level.

```{r volcano_DEgenes,eval=T,echo=F,fig.width=12}
fig2insert <- paste0(DDD.data$figure.folder,'/DE genes volcano plot.png')
if(!file.exists(fig2insert)){
  cat('Plot is not found in the figure folder')
} else {
  knitr::include_graphics(fig2insert)
}
```


```{r volcano_DASgenes,eval=T,echo=F,fig.width=12}
fig2insert <- paste0(DDD.data$figure.folder,'/DAS genes volcano plot.png')
if(!file.exists(fig2insert)){
  cat('Plot is not found in the figure folder')
} else {
  knitr::include_graphics(fig2insert)
}
```

```{r volcano_DEtrans,eval=T,echo=F,fig.width=12}
fig2insert <- paste0(DDD.data$figure.folder,'/DE transcripts volcano plot.png')
if(!file.exists(fig2insert)){
  cat('Plot is not found in the figure folder')
} else {
  knitr::include_graphics(fig2insert)
}
```


```{r volcano_DTUtrans,eval=T,echo=F,fig.width=12}
fig2insert <- paste0(DDD.data$figure.folder,'/DTU transcripts volcano plot.png')
if(!file.exists(fig2insert)){
  cat('Plot is not found in the figure folder')
} else {
  knitr::include_graphics(fig2insert)
}
```

### Heatmap
Hierarchical clustering was used to partition the DE genes into `r params$DDD.data$params_list$cluster_number` clusters with `r params$DDD.data$params_list$dist_method` distance and `r params$DDD.data$params_list$cluster_method` clustering algorithm (Saracli et al., 2013). ComplexHeatmap R package version 1.20.0 was used to make the heat-maps.

Figure: Heatmap of DE genes, including DE only and DE+DAS genes. 

```{r Heatmap_DEgenes,eval=T,echo=F,fig.width=6,fig.height=8}
## DE genes
fig2insert <- paste0(DDD.data$figure.folder,'/Heatmap DE genes.png')
if(!file.exists(fig2insert)){
  cat('Plot is not found in the figure folder')
} else {
  knitr::include_graphics(fig2insert)
}
```


Figure: Heatmap of DE+DAS genes. The genes of DAS only were removed since they only had significant alternative splicing changes, but did not have significant expression changes. 

```{r Heatmap_DASgenes,eval=T,echo=F,fig.width=6,fig.height=8}
## DAS genes
fig2insert <- paste0(DDD.data$figure.folder,'/Heatmap DAS genes.png')
if(!file.exists(fig2insert)){
  cat('Plot is not found in the figure folder')
} else {
  knitr::include_graphics(fig2insert)
}
```


Figure: Heatmap of DE transcripts, including DE only and DE+DTU transcripts.

```{r Heatmap_DEtrans,eval=T,echo=F,fig.width=6,fig.height=8}
## DE transcripts
fig2insert <- paste0(DDD.data$figure.folder,'/Heatmap DE transcripts.png')
if(!file.exists(fig2insert)){
  cat('Plot is not found in the figure folder')
} else {
  knitr::include_graphics(fig2insert)
}
```

Figure: Heatmap of DE+DTU transcripts.DTU only transcripts were removed since they only had significant alternative splicing changes, but did not have significant expression changes.

```{r Heatmap_DTUtrans,eval=T,echo=F,fig.width=6,fig.height=8}
fig2insert <- paste0(DDD.data$figure.folder,'/Heatmap DTU transcripts.png')
## DTU transcripts
if(!file.exists(fig2insert)){
  cat('Plot is not found in the figure folder')
} else {
  knitr::include_graphics(fig2insert)
}
```

### GO annotation plot

Figure: Significant GO enriched terms of DE genes.

```{r GO_DEgenes,eval=T,echo=F,fig.width=12}
fig2insert <- paste0(DDD.data$figure.folder,'/DE genes GO annotation plot.png')
if(!file.exists(fig2insert)){
  cat('Plot is not found in the figure folder')
} else {
  knitr::include_graphics(fig2insert)
}
```

<br>

Figure: Significant GO enriched terms of DAS genes.

```{r GO_DASgenes,eval=T,echo=F,fig.width=12}
fig2insert <- paste0(DDD.data$figure.folder,'/DAS genes GO annotation plot.png')
if(!file.exists(fig2insert)){
  cat('Plot is not found in the figure folder')
} else {
  knitr::include_graphics(fig2insert)
}
```



### Significant isoform switches

Figure: Significant isoform switches between transcript isoforms.

```{r ISs,eval=T,echo=F,fig.width=12}
fig2insert <- paste0(DDD.data$figure.folder,'/Isoform switch number.png')
if(!file.exists(fig2insert)){
  cat('Plot is not found in the figure folder')
} else {
  knitr::include_graphics(fig2insert)
}
```


## Supplementary figures
### Mean-variance trend plot
The cut-offs to filter the transcripts were determined by the mean-variance trend plots (Law et al. 2014).

- An expressed transcript must have $\geq$ `r params$DDD.data$params_list$cpm_samples_n` out of 9 samples with CPM $\geq$ `r params$DDD.data$params_list$cpm_cut`.
- An expressed gene must have at least one expressed transcript.

Figure: Transcript level mean-variance trend plot

```{r,eval=T,echo=F,fig.width=12,message=F,warning=F}
fig2insert <- paste0(DDD.data$figure.folder,'/Transcript mean-variance trend.png')
if(!file.exists(fig2insert)){
  cat('Plot is not found in the figure folder')
} else {
  knitr::include_graphics(fig2insert)
}
```

Figure: Gene level mean-variance trend plot

```{r,eval=T,echo=F,fig.width=12,message=F,warning=F}
fig2insert <- paste0(DDD.data$figure.folder,'/Gene mean-variance trend.png')
if(!file.exists(fig2insert)){
  cat('Plot is not found in the figure folder')
} else {
  knitr::include_graphics(fig2insert)
}
```

### PCA plot of all samples
The normalised $log_2$-CPM of all samples were used to make PCA plots. The PCA plot of all samples can be found in the figure folder.


### Sample distribution

Figure: Transcript level read counts and normalised $log_2$-CPM distribution across samples.

```{r,eval=T,echo=F,fig.width=12,message=F,warning=F}
fig2insert <- paste0(DDD.data$figure.folder,'/Transcript expression distribution.png')
if(!file.exists(fig2insert)){
  cat('Plot is not found in the figure folder')
} else {
  knitr::include_graphics(fig2insert)
}

```

Figure: Gene level read counts and normalised $log_2$-CPM distribution across samples.

```{r,eval=T,echo=F,fig.width=12,message=F,warning=F}
fig2insert <- paste0(DDD.data$figure.folder,'/Gene expression distribution.png')
if(!file.exists(fig2insert)){
  cat('Plot is not found in the figure folder')
} else {
  knitr::include_graphics(fig2insert)
}
```

### Venn diagrams
Venn diagrams to compare significant DE genes, DAS genes, DE transcript and DTU transcripts in different contrast groups can be found in figure folder.

## Supplementary materials

### Files in report folder
Reports are saved in report folder.

| File name              | Description                                       |
|------------------------|---------------------------------------------------|
| 3D_report.pdf/html/doc | Report of 3D analysis in pdf, html and doc format |


### Files in figure folder

|File.names                                            |Description                                  |
|:-----------------------------------------------------|:--------------------------------------------|
|DAS genes GO annotation plot.png/.pdf                 |DAS genes GO annotation plot                 |
|DAS genes updown regulation numbers.png/.pdf          |DAS genes up-down regulation numbers          |
|DAS genes volcano plot.png/.pdf                       |DAS genes volcano plot                       |
|DE genes GO annotation plot.png/.pdf                  |DE genes GO annotation plot                  |
|DE genes updown regulation numbers.png/.pdf           |DE genes up-down regulation numbers           |
|DE genes volcano plot.png/.pdf                        |DE genes volcano plot                       |
|DE transcripts updown regulation numbers.png/.pdf     |DE transcripts up-down regulation numbers     |
|DE transcripts volcano plot.png/.pdf                  |DE transcripts volcano plot                       |
|DTU transcripts updown regulation numbers.png/.pdf    |DTU transcripts up-down regulation numbers    |
|DTU transcripts volcano plot.png/.pdf                 |DTU transcripts volcano plot                       |
|Gene data distribution.png/.pdf                       |Gene data distribution                       |
|Gene mean-variance trend.png/.pdf                     |Gene mean-variance trend                     |
|Gene PCA Average expression.png/.pdf                  |Gene PCA Average expression                  |
|Gene PCA batch effect removed Bio-reps.png/.pdf       |Gene PCA batch effect removed Bio-reps       |
|Gene PCA Bio-reps.png/.pdf                            |Gene PCA Bio-reps                            |
|Heatmap DAS genes.png/.pdf                            |Heatmap of DE+DAS genes                            |
|Heatmap DE genes.png/.pdf                             |Heatmap of DE genes                             |
|Heatmap DE transcripts.png/.pdf                       |Heatmap of DE transcripts                       |
|Heatmap DTU transcripts.png/.pdf                      |Heatmap of DE+DTU transcripts                      |
|Transcript data distribution.png/.pdf                 |Transcript data distribution                 |
|Transcript mean-variance trend.png/.pdf               |Transcript mean-variance trend               |
|Transcript PCA Average expression.png/.pdf            |Transcript PCA Average expression            |
|Transcript PCA batch effect removed Bio-reps.png/.pdf |Transcript PCA batch effect removed Bio-reps |
|Transcript PCA Bio-reps.png/.pdf                      |Transcript PCA Bio-reps                      |
|Union set DE genes vs DAS genes.png/.pdf              |Flow chart -Union set DE genes vs DAS genes              |
|Union set DE transcripts vs DTU transcripts.png/.pdf  |Flow chart -Union set DE transcripts vs DTU transcripts  |
|Isoform switch number.png/.pdf                        |Number of significant isoform switch numbers in contrast groups|

### Files in result folder
Important results are saved in csv (comma delimited) files.

| csv files                                                       | Description                                                                                                                                                              |   |
|-----------------------------------------------------------------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------|---|
| contrast.csv                                                    | Contrast groups used for 3D analysis                                                                                                                                     |   |
| DDD genes and transcript lists across all   contrast groups.csv | List of DE genes, DAS genes, DE transcripts and DTU transcripts, which   are the union sets across all contrast groups                                                   |   |
| DDD numbers.csv                                                 | DE/DAS/DTU genes/transcript numbers in each contrast group                                                                                                               |   |
| DEvsDAS results.csv                                             | Number of DE vs DAS genes                                                                                                                                                |   |
| DEvsDTU results.csv                                             | Number of DE vs DTU transcripts                                                                                                                                          |   |
| Gene read counts.csv                                            | Raw read counts of genes before data pre-processing                                                                                                                      |   |
| Gene TPM.csv                                                    | Raw TPM of genes before data pre-processing                                                                                                                              |   |
| RNAseq info.csv                                                 | RNA-seq data information before   and after pre-processing                                                                                                               |   |
| samples.csv                                                     | Sample information                                                                                                                                                       |   |
| Significant DE genes list and   statistics.csv                  | Statistics of significant DE genes                                                                                                                                       |   |
| Significant DE genes list and   statistics.csv                  | Statistics of significant DAS genes                                                                                                                                      |   |
| Significant DE transcripts list and   statistics.csv            | Statistics of significant DE transcripts                                                                                                                                 |   |
| Significant DTU transcripts list and   statistics.csv           | Statistics of significant DTU transcripts                                                                                                                                |   |
| Target in each cluster heatmap*DE   genes.csv                   | DE gene list in clusters of DE gene heatmap.                                                                                                                             |   |
| Target in each cluster heatmap*DE   trans.csv                   | DE&DTU transcript lists in clusters of DTU transcript heatmap. The   DTUonly transcripts are excluded since they have no significant abundance   changes across samples. |   |
| Target in each cluster heatmap*DE&DAS   genes.csv               | DE&DAS gene lists in clusters of DAS gene heatmap. The DASonly genes   are excluded since they have no significant abundance changes across samples.                     |   |
| Target in each cluster heatmap*DE&DTU   trans.csv               | DE transcript list in clusters of DE transcript heatmap.                                                                                                                 |   |
| Transcript read counts.csv                                      | Raw read counts of transcripts before data pre-processing                                                                                                                |   |
| Transcript TPM.csv                                              | Raw TPM of transcripts before data pre-processing                                                                                                                        |   |

### Files in data folder
Intermediate data in .RData for 3D RAN-seq analysis are saved in the data folder. There are three .RData objects: 1) txi_trans.RData and 2) txi_genes.RData are transcript and gene level read count and TPM outputs from the tximport R package (Soneson et al., 2016). All the intermediate data generated in the process of 3D analysis is saved in the list object intermediate_data.RData. R users can access to the data using command line. 

| List   object                   | Elements in list object | Element type | Description                                                                                                                                                                                                                                                                                          |
|---------------------------------|-------------------------|--------------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| intermediate_data.RData         | conditions              | character    | Labels of conditions to study                                                                                                                                                                                                                                                                                |
|                                 | contrast                | character    | Contrast groups                                                                                                                                                                                                                                                                                      |
|                                 | DAS_genes               | data.frame   | Statistics of significant DTU transcripts                                                                                                                                                                                                                                                            |
|                                 | DDD_numbers             | data.frame   | Number of DE/DAS/DTU genes/transcripts in contrast groups                                                                                                                                                                                                                                            |
|                                 | DE_genes                | data.frame   | Statistics of significant DE genes                                                                                                                                                                                                                                                                   |
|                                 | DE_trans                | data.frame   | Statistics of significant DE transcripts                                                                                                                                                                                                                                                             |
|                                 | deltaPS                 | data.frame   | Delta PS based on contrast groups                                                                                                                                                                                                                                                                    |
|                                 | DEvsDAS_results         | data.frame   | Number of DE vs DAS genes                                                                                                                                                                                                                                                                            |
|                                 | DEvsDTU_results         | data.frame   | Number of DE vs DTU transcripts                                                                                                                                                                                                                                                                      |
|                                 | DTU_trans               | data.frame   | Statistics of significant DTU transcripts                                                                                                                                                                                                                                                            |
|                                 | genes_3D_stat           | list         | All the raw results of linear regression and statistics of DE   genes                                                                                                                                                                                                                                |
|                                 | genes_batch             | list         | Estimated gene level batch effects, if they exist. 1) W: matrix,   estimated batch effect term, which can be added to design matrix of linear   regression; 2) normalizedCounts: matrix, read counts where batch effects are   removed; 3) method: a string, method used to estimate batch effects.         |
|                                 | genes_counts            | data.frame   | Read counts of genes. Seq-reps are merged if exist.                                                                                                                                                                                                                                                  |
|                                 | genes_log2FC            | matrix       | log2-CPM of genes                                                                                                                                                                                                                                                                                    |
|                                 | genes_TPM               | matrix       | TPMs of genes                                                                                                                                                                                                                                                                                 |
|                                 | mapping                 | data.frame   | Transcript-gene mapping                                                                                                                                                                                                                                                                              |
|                                 | params_list             | list         | Parameters used for the 3D analysis                                                                                                                                                                                                                                                                  |
|                                 | PS                      | matrix       | Percent spliced (PS) of expressed transcripts                                                                                                                                                                                                                                                        |
|                                 | RNAseq_info             | data.frame   | RNA-seq data information before and after pre-processing                                                                                                                                                                                                                                             |
|                                 | samples                 | data.frame   | Sample information.                                                                                                                                                                                                                                                                                  |
|                                 | samples_new             | data.frame   | Sample information after merging sequencing replicates   (seq-reps, if exist).                                                                                                                                                                                                                       |
|                                 | scores                  | data.frame   | Statistics of isoform switches                                                                                                                                                                                                                                                                       |
|                                 | scores_filtered         | data.frame   | Statistics of significant isoform switches                                                                                                                                                                                                                                                           |
|                                 | target_high             | list         | 1) trans_high: character, expressed transcripts; 2)   genes_high: character, expressed genes; 3) mapping_high: data.frame,   expressed transcript-gene mapping                                                                                                                                       |
|                                 | trans_3D_stat           | list         | All the raw results of linear regression and statistics of DAS   genes, DE and DTU transcripts                                                                                                                                                                                                       |
|                                 | trans_batch             | list         | Estimated transcript level batch effects, if they exist. 1) W:   matrix, estimated batch effect term, which can be added to design matrix of   linear regression; 2) normalizedCounts: matrix, read counts where batch   effects are removed; 3) method: string, method used to estimate batch   effects. |
|                                 | trans_counts            | data.frame   | Read counts of transcripts. Seq-reps are merged if exist.                                                                                                                                                                                                                                            |
|                                 | trans_log2FC            | matrix       | log2-CPM of transcripts.                                                                                                                                                                                                                                                                             |
|                                 | trans_TPM               | matrix       | TPMs of transcripts.                                                                                                                                                                                                                                                                                 |
|                                 | Other elements          |              | The list object may include other elements.                                                                                                                                                                                                                                                          |
| txi_genes.Rdata/txi_trans.Rdata | abundance               | matrix       | TPMs of genes/transcripts                                                                                                                                                                                                                                                                            |
|                                 | counts                  | matrix       | Read counts of genes/transcripts                                                                                                                                                                                                                                                                     |
|                                 | countsFromAbundance     | character    | Method used to generate read counts and TPMs                                                                                                                                                                                                                                                         |
|                                 | length                  | matrix       | Length of genes/transcripts                                                                                                                                                                                                                                                                          |                                                                                                                                                                                                                        |


## References
Benjamini,Y. and Yekutieli,D. (2001) The control of the false discovery rate in multiple testing under dependency. Ann. Stat., 29, 1165–1188.

Bray,N.L., Pimentel,H., Melsted,P., and Pachter,L. (2016) Near-optimal probabilistic RNA-seq quantification. Nat. Biotechnol., 34, 525–527.

Bullard,J.H., Purdom,E., Hansen,K.D., and Dudoit,S. (2010) Evaluation of statistical methods for normalization and differential expression in mRNA-Seq experiments. BMC Bioinformatics, 11, 94.

Calixto,C.P.G., Guo,W., James,A.B., Tzioutziou,N.A., Entizne,J.C., Panter,P.E., Knight,H., Nimmo,H., Zhang,R., and Brown,J.W.S. (2018) Rapid and dynamic alternative splicing impacts the Arabidopsis cold response transcriptome. Plant Cell, tpc.00177.2018.

Gu,Z., Eils,R., and Schlesner,M. (2016) Complex heatmaps reveal patterns and correlations in multidimensional genomic data. Bioinformatics, 32, 2847–2849.

Guo,W., Calixto,C.P.G., Brown,J.W.S., and Zhang,R. (2017) TSIS: An R package to infer alternative splicing isoform switches for time-series data. Bioinformatics, 33, 3308–3310.

Guo,W., Tzioutziou,N., Stephen,G., Milne,I., Calixto,C., Waugh,R., Brown,J.W., and Zhang,R. (2019) 3D RNA-seq - a powerful and flexible tool for rapid and accurate differential expression and alternative splicing analysis of RNA-seq data for biologists. bioRxiv, 656686. doi: https://doi.org/10.1101/656686.

Law,C.W., Chen,Y., Shi,W., and Smyth,G.K. (2014) voom: Precision weights unlock linear model analysis tools for RNA-seq read counts. Genome Biol, 15, R29.

Patro,R., Duggal,G., Love,M.I., Irizarry,R.A., and Kingsford,C. (2017) Salmon provides fast and bias-aware quantification of transcript expression. Nat. Methods, 14, 417–419.

Risso,D., Ngai,J., Speed,T.P., and Dudoit,S. (2014) Normalization of RNA-seq data using factor analysis of control genes or samples. Nat. Biotechnol., 32, 896–902.

Ritchie,M.E., Phipson,B., Wu,D., Hu,Y., Law,C.W., Shi,W., and Smyth,G.K. (2015) limma powers differential expression analyses for RNA-sequencing and microarray studies. Nucleic Acids Res, 43, e47.

Saraçli,S., Dogan,N., and Dogan,I. (2013) Comparison of hierarchical cluster analysis methods by cophenetic correlation. J. Inequalities Appl.

Soneson,C., Love,M.I., and Robinson,M.D. (2016) Differential analyses for RNA-seq: transcript-level estimates improve gene-level inferences. F1000Research, 4, 1521.


## Session information

```{r,echo=F,eval=TRUE}
sessionInfo()
```


